lib/seahorn/PredicateAbstraction.cc

```diff
@@ -0,0 +1,582 @@
+					//converter
+					boolToTermMap.insert(std::pair<Expr, Expr>(bind::bvar(index, mk<BOOL_TY>(term_app->efac())), term));
+
+					new_body_exprs.push_back(equal_expr);
+					index ++;
+				}
+				converter.addRelToBoolToTerm(bind::fname(*it), boolToTermMap);
+				//converter.getRelToBoolToTermMap().insert(std::pair<Expr, ExprMap>(bind::fname(*it), boolToTermMap));
+			}
+
+			Expr rule_head = r.head();
+
+			//construct new rule head.
+			Expr new_rule_head_rel = m_oldToNewPredMap.find(bind::fname(rule_head))->second;
+
+			ExprVector new_rule_head_args;
+			int pred_order = relOccurrenceTimesMap.find(bind::fname(rule_head))->second;
+			relOccurrenceTimesMap[bind::fname(rule_head)] += 1;
+			for(int i=0; i<bind::domainSz(new_rule_head_rel); i++)
+			{
+				Expr var_tag = variant::variant(pred_order, variant::variant(i, variant::tag(bind::fname(new_rule_head_rel), mkTerm<std::string> ("p", new_rule_head_rel->efac ())))); //prime
+				Expr boolVar = bind::boolConst(var_tag);
+				rule_vars.push_back(boolVar);
+				new_rule_head_args.push_back(boolVar);
+			}
+
+			Expr new_rule_head = bind::fapp(new_rule_head_rel, new_rule_head_args);
+			LOG("pabs-debug", outs() << "NEW RULE HEAD: " << *new_rule_head << "\n";);
+
+
+			if(bind::domainSz(bind::fname(rule_head)) != 0)
+			{
+				//construct head equality expr, put in new body
+				int index = 0;
+				//for converter
+				ExprMap boolToTermMap;
+				for(Expr term : m_currentCandidates.find(bind::fname(rule_head))->second)
+				{
+					Expr term_app = util_applyArgsToBvars(term, rule_head, m_currentCandidates);
+					Expr equal_expr = mk<IFF>(new_rule_head->arg(index + 1), term_app);
+					//converter
+					boolToTermMap.insert(std::pair<Expr, Expr>(bind::bvar(index, mk<BOOL_TY>(term_app->efac())), term));
+
+					new_body_exprs.push_back(equal_expr);
+					index ++;
+				}
+				converter.addRelToBoolToTerm(bind::fname(rule_head), boolToTermMap);
+				//converter.getRelToBoolToTermMap().insert(std::pair<Expr, ExprMap>(bind::fname(rule_head), boolToTermMap));
+			}
+
+			//Extract the constraints
+			Expr constraints = util_extractTransitionRelation(r, db);
+			new_body_exprs.push_back(constraints);
+
+			//Construct new body
+			Expr new_rule_body = mknary<AND>(new_body_exprs.begin(), new_body_exprs.end());
+			LOG("pabs-debug", outs() << "NEW RULE BODY :" << *new_rule_body << "\n";);
+
+			HornRule new_rule(rule_vars, new_rule_head, new_rule_body);
+			new_DB.addRule(new_rule);
+		}
+	}
+
+	void PredicateAbstractionAnalysis::generateAbstractQueries(HornClauseDB &db, HornClauseDB &new_DB)
+	{
+		for(auto old_query : db.getQueries())
+		{
+			Expr old_fdecl = bind::fname(old_query);
+			Expr new_fdecl = m_oldToNewPredMap.find(old_fdecl)->second;
+			ExprVector old_args;
+			Expr new_query = bind::fapp(new_fdecl, old_args);
+			new_DB.addQuery(new_query);
+		}
+	}
+
+	void PredicateAbstractionAnalysis::guessCandidate(HornClauseDB &db)
```

Comment:
should not guessing of terms be factored out and shared between PredAbs and Houdini?